api/v1alpha2/jenkins_webhook.go

```diff
@@ -37,25 +56,328 @@ func (in *Jenkins) SetupWebhookWithManager(mgr ctrl.Manager) error {
+					lastVersion := version.LastVersion
+					if len(firstVersion) == 0 {
+						firstVersion = "0" // setting default value in case of empty string
+					}
+					if len(lastVersion) == 0 {
+						lastVersion = pluginData.Version // setting default value in case of empty string
+					}
+
+					if compareVersions(firstVersion, lastVersion, pluginData.Version) {
+						jenkinslog.Info("Security Vulnerability detected in "+pluginData.Kind+" "+plugin.Name+":"+pluginData.Version, "Warning message", warning.Message, "For more details,check security advisory", warning.URL)
+						hasVulnerabilities = true
+					}
+				}
+			}
+
+			if hasVulnerabilities {
+				if pluginData.Kind == "base" {
+					faultyBasePlugins += "\n" + plugin.Name + ":" + pluginData.Version
+				} else {
+					faultyUserPlugins += "\n" + plugin.Name + ":" + pluginData.Version
+				}
+			}
+		}
+	}
+	if len(faultyBasePlugins) > 0 || len(faultyUserPlugins) > 0 {
+		var errormsg string
+		if len(faultyBasePlugins) > 0 {
+			errormsg += "security vulnerabilities detected in the following base plugins: " + faultyBasePlugins
+		}
+		if len(faultyUserPlugins) > 0 {
+			errormsg += "security vulnerabilities detected in the following user-defined plugins: " + faultyUserPlugins
+		}
+		return errors.New(errormsg)
+	}
+
+	return nil
+}
+
+func NewPluginsDataManager(hosturl string, compressedFilePath string, pluginDataFile string, isCached bool, timeout time.Duration) *PluginDataManager {
+	return &PluginDataManager{
+		Hosturl:            hosturl,
+		CompressedFilePath: compressedFilePath,
+		PluginDataFile:     pluginDataFile,
+		IsCached:           isCached,
+		Timeout:            timeout,
+	}
+}
+
+// Downloads extracts and reads the JSON data in every 12 hours
+func (in *PluginDataManager) FetchPluginData(isInitialized chan bool) {
+	for {
+		jenkinslog.Info("Initializing/Updating the plugin data cache")
+		var isDownloaded, isExtracted, isCached bool
+		var err error
+		for in.Attempts = 0; in.Attempts < 5; in.Attempts++ {
+			err = in.download()
+			if err == nil {
+				isDownloaded = true
+				break
+			}
+		}
+
+		if isDownloaded {
+			for in.Attempts = 0; in.Attempts < 5; in.Attempts++ {
+				err = in.extract()
+				if err == nil {
+					isExtracted = true
+					break
+				}
+			}
+		} else {
+			jenkinslog.Info("Cache Plugin Data", "failed to download file", err)
+		}
+
+		if isExtracted {
+			for in.Attempts = 0; in.Attempts < 5; in.Attempts++ {
+				err = in.cache()
+				if err == nil {
+					isCached = true
+					break
+				}
+			}
+
+			if !isCached {
+				jenkinslog.Info("Cache Plugin Data", "failed to read plugin data file", err)
+			}
+		} else {
+			jenkinslog.Info("Cache Plugin Data", "failed to extract file", err)
+		}
+
+		// Checks for the first time
+		if !in.IsCached {
+			isInitialized <- isCached
+			in.IsCached = isCached
+		}
+
+		if isCached {
+			in.SleepTime = 12
+		} else {
+			in.SleepTime = 1
+		}
+		time.Sleep(time.Duration(in.SleepTime) * time.Hour)
```

Comment:
```suggestion
		if !in.IsCached {
			isInitialized <- isCached
			in.IsCached = isCached
			in.sleepTime = 1
		}
		
		in.sleepTime = 12
		time.Sleep(in.sleepTime)

```

Reply:
in.IsCached defines that whether we have cached the data file or not,it will always be true once we successfully cached for the first time, isCached defines our current case, so if we are updating our cache and it fails we still are going to run the function after 1 hour.